lib/discover_endpoint.js

```diff
@@ -0,0 +1,340 @@
+var AWS = require('./core');
+var util = require('./util');
+var ENDPOINT_OPERATION_MAX_RETRIES = 60;
+var endpointDiscoveryEnabledEnv = 'AWS_ENABLE_ENDPOINT_DISCOVERY';
+
+/**
+ * Generate key to index the endpoints in the cache
+ * @return [map<String,String>] object with keys to index endpoints.
+ * @api private
+ */
+function getCacheKey(request) {
+  var service = request.service;
+  var api = service.api || {};
+  var operations = api.operations;
+  var identifiers = {};
+  if (operations[request.operation] && operations[request.operation].name) {
```

Comment:
I think operation is only required if custom identifiers are defined for an operation. Is that not the case? Might help to keep the size of your cache down if we omitted operation if it isn't needed.